### ISSUE TITLE
Consider adding the ability to register custom loggers

### ISSUE THREAD

**Issue by canova** [rayon-rs/rayon#915]
Hi everyone, in Firefox, we were trying to record and mark the thread sleeps with the Firefox Profiler. That way, we can profile and see when a thread is starting to sleep and when it's getting awake. It's pretty easy to mark these sleeps in the C++ side since we handle these sleeps ourselves there. But it's more tricky for the Rust threads, because mostly we are using Rayon to handle these and we don't have an access to sleep code path.

So I tried to find a handler that could be called right before the sleep and right after the sleep, but I can't find an API like that on Rayon. I guess, having this API could be also useful, but it deserves its own issue (if you think it would be valuable). While I was looking at the Rayon codebase, I saw that there are some logs right [before](https://github.com/rayon-rs/rayon/blob/f45eee8fa49c1646a00f084ca78d362f381f1b65/rayon-core/src/sleep/mod.rs#L198-L201) and [after](https://github.com/rayon-rs/rayon/blob/f45eee8fa49c1646a00f084ca78d362f381f1b65/rayon-core/src/sleep/mod.rs#L234-L237) the sleep. At that moment I was kinda excited thinking that I could maybe register my own custom logger and get those notifications. But unfortunetly, [I don't see](https://github.com/rayon-rs/rayon/blob/f45eee8fa49c1646a00f084ca78d362f381f1b65/rayon-core/src/log.rs#L1-L19) any custom logging features right now in the codebase.

So, I would really love if we could have a Logger trait that people could implement and register that as a custom logger while we are building a thread pool ([for example here](https://searchfox.org/mozilla-central/rev/b697834e78a3ef7613e2fa57c07624b1d9d1c909/gfx/webrender_bindings/src/bindings.rs#1096-1110)). That way, we can have access to the thread sleeps and we can add gecko profiler markers to those. Optionally, we are thinking about adding some telemetry, so we can understand more and optimize these use cases.

**Comment by nikomatsakis** [comment]
I think what I would like to do is to integrate with `tracing`-- do you think that would suit your needs, @canova ?

**Comment by canova** [comment]
@nikomatsakis Yes, probably that would be a more generic way to do this. Currently we don't have a way to collect `tracing` events in the Firefox Profiler side but we do [have a bug filed for it](https://bugzilla.mozilla.org/show_bug.cgi?id=1652558). That would indeed suit us well.

**Comment by nikomatsakis** [comment]
@cuviper yes, the 0.1 status is a blocker for adopting tracing.

**Comment by RReverser** [comment]
Heh, I was about to file a feature request for `tracing` integration but then found this issue. I'll post my repro + current workaround anyway:

```rust
use rayon::prelude::*;
use tracing::{info, info_span, instrument};
use tracing_subscriber::prelude::*;
use tracing_tree::HierarchicalLayer;

#[instrument]
fn do_smth(value: i32) -> i32 {
    value
}

fn main() {
    let subscriber = tracing_subscriber::Registry::default().with(HierarchicalLayer::default());
    tracing::subscriber::set_global_default(subscriber).unwrap();

    info_span!("iter").in_scope(|| {
        info!(sum = vec![1, 2, 3, 4].into_iter().map(do_smth).sum::<i32>());
    });

    info_span!("rayon").in_scope(|| {
        info!(sum = vec![1, 2, 3, 4].into_par_iter().map(do_smth).sum::<i32>());
    });

    let explicit_span = info_span!("rayon with explicit parent");
    explicit_span.in_scope(|| {
        info!(
            sum = vec![1, 2, 3, 4]
                .into_par_iter()
                .map(|value| {
                    let _explicit_span_enter = explicit_span.enter();
                    do_smth(value)
                })
                .sum::<i32>()
        );
    });
}
```

Output:

```
iter 
  do_smth value=1

  do_smth value=2

  do_smth value=3

  do_smth value=4

  2ms  INFO sum=10

rayon
do_smth value=3

do_smth value=2

do_smth value=4

do_smth value=1

  3ms  INFO sum=10

rayon with explicit parent
  do_smth value=1

  do_smth value=4

  do_smth value=3

  do_smth value=2

  4ms  INFO sum=10
```

It shows that the problem is that tracing doesn't know the parents of spans created in new threads. It's possible to work around this manually by explicitly entering parent spans in every iterator callback, but it gets very cumbersome.

I think merging

> [acquire_thread_handler](https://docs.rs/rustc-rayon/latest/rustc_rayon/struct.ThreadPoolBuilder.html#method.acquire_thread_handler) and [release_thread_handler](https://docs.rs/rustc-rayon/latest/rustc_rayon/struct.ThreadPoolBuilder.html#method.release_thread_handler)

to mainstream of Rayon would help address this issue nicely without having to depend on any particular tracing library. Is it a lot of work to merge it / were there any blockers in the past?

**Comment by RReverser** [comment]
> I think merging
> 
> > [acquire_thread_handler](https://docs.rs/rustc-rayon/latest/rustc_rayon/struct.ThreadPoolBuilder.html#method.acquire_thread_handler) and [release_thread_handler](https://docs.rs/rustc-rayon/latest/rustc_rayon/struct.ThreadPoolBuilder.html#method.release_thread_handler)
> 
> to mainstream of Rayon would help address this issue nicely without having to depend on any particular tracing library.

Although, looking at the signatures, I'm not sure how span entering would work with those callbacks. We need to be able to 1) acquire a span on the caller thread 2) enter that span inside the task on a separate thread. The callbacks seem to allow only part of this?

**Comment by cuviper** [comment]
Yeah, I'm not sure how to support contextual spans with those callbacks. The `rustc-rayon` use case is only for acquiring and releasing jobserver tokens, as a global inter-process resource.

**Comment by RReverser** [comment]
I suppose it would need two callbacks - one to "prepare" on the main thread and possibly return a value (`FnOnce() -> T where T: Sync`), and another that's actually executed before the task on the pool thread (taking `&T` as an argument).

**Comment by wagnerf42** [comment]
note that there is another workaround:
you can have a special iterator adaptor which does the parent registration.
i have some logs using tracing in rayon here :
https://github.com/wagnerf42/diam/blob/main/examples/log.rs

**Comment by RReverser** [comment]
[Update a few months later.] Despite being marked as 0.1, I see tracing is becoming ubiquitous enough in the ecosystem to warrant including its support in rayon IMO, at least as an optional dependency.

**Comment by RReverser** [comment]
@cuviper Would you be open to revisiting this yet? tracing became very popular over those years and inability to use it transparently with Rayon is painful sometimes :(
